relayapi/storage/shared/storage.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+
+	"github.com/matrix-org/dendrite/federationapi/storage/shared"
+	"github.com/matrix-org/dendrite/internal/caching"
+	"github.com/matrix-org/dendrite/internal/sqlutil"
+	"github.com/matrix-org/dendrite/relayapi/storage/tables"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+type Database struct {
+	DB                *sql.DB
+	IsLocalServerName func(gomatrixserverlib.ServerName) bool
+	Cache             caching.FederationCache
+	Writer            sqlutil.Writer
+	RelayQueue        tables.RelayQueue
+	RelayQueueJSON    tables.RelayQueueJSON
+}
+
+func (d *Database) StoreAsyncTransaction(
+	ctx context.Context, txn gomatrixserverlib.Transaction,
+) (*shared.Receipt, error) {
+	var err error
+	json, err := json.Marshal(txn)
+	if err != nil {
+		return nil, fmt.Errorf("d.JSONUnmarshall: %w", err)
+	}
+
+	var nid int64
+	_ = d.Writer.Do(d.DB, nil, func(txn *sql.Tx) error {
+		nid, err = d.RelayQueueJSON.InsertQueueJSON(ctx, txn, string(json))
+		return err
+	})
+	if err != nil {
+		return nil, fmt.Errorf("d.insertQueueJSON: %w", err)
+	}
+
+	receipt := shared.NewReceipt(nid)
+	return &receipt, nil
+}
+
+func (d *Database) AssociateAsyncTransactionWithDestinations(
+	ctx context.Context,
+	destinations map[gomatrixserverlib.UserID]struct{},
+	transactionID gomatrixserverlib.TransactionID,
+	receipt *shared.Receipt,
+) error {
+	for destination := range destinations {
+		err := d.Writer.Do(d.DB, nil, func(txn *sql.Tx) error {
+			err := d.RelayQueue.InsertQueueEntry(
+				ctx, txn, transactionID, destination.Domain(), receipt.GetNID())
```

Comment:
It would be better to do this in bulk.

As it stands this is also buggy because a new txn is made for each insert, so we can end up partially inserting destinations.